### PR TITLE
ES2015 and React

### PR DESCRIPTION
**Pull request by nii236**
Mostly a dump of linter rules plus a few more dot points

### CODE REVIEW

File: js-es2015-react.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+The `.eslintignore` file contains only `node_modules` and whatever project specific folders as required.
+
+# Styleguide
+
+General points:
+
+- Classes are PascalCased
+- Functions are camelCased
+- React Props and States are PascalCased
```
Comment (osiro) [inline js-es2015-react.md]: That's unusual...

React Props and States are usually camelCased

File: js-es2015-react.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
+# Tooling
+
+For projects that use ES2015, we use:
+
+- eslint
+- eslint-plugin-react
+
+To lint our javascript files.
```
Comment (sevenseacat) [inline js-es2015-react.md]: the name of the language is also "PascalCased" (ie. JavaScript)

File: js-es2015-react.md (hunk 1 above)
Comment (sevenseacat) [inline js-es2015-react.md]: +1

Comment (nii236) [inline js-es2015-react.md]: Good catch!

**Comment by osiro** [comment]
@nii236 do you think we should also mention the usage of `propTypes` as part of the standards?

@pzi any preference/suggestion about the order of lifecycle methods? Should them be defined in the end of components?

Also, what you guys think about the details below?

# File name

Considering we often use ReactJS with webpack in Rails apps, I believe it's convenient follow the Rails file names, for instance: `/webpack/components/upcoming_deliveries.jsx` instead of `/webpack/components/UpcomingDeliveries.jsx`.

# Separation

I think we need more separation.

In WorkReady we have `App.jsx` but in a Rails app we can have many different mini ReactJS apps that are not related to each other.

screen shot 2015-11-19 at 12 16 17 pm

I would separate them in main mini apps, like:

screen shot 2015-11-19 at 12 17 24 pm

screen shot 2015-11-19 at 12 17 35 pm

So, if for whatever reason ElMulcho requires another ReactJS app for something nothing to do with `delivery`, I can just create another `app.jsx` under the folder `orders` for example.

**Comment by levibuzolic** [comment]
> Considering we often use ReactJS with webpack in Rails apps, I believe it's convenient follow the Rails file names, for instance: /webpack/components/upcoming_deliveries.jsx instead of /webpack/components/UpcomingDeliveries.jsx.

The `_something_here.jsx` pattern is Sprockets specific to solve the automatic importing of directories, and I wouldn't recommend using it with Webpack because it doesn't do anything. :ok_hand: 

Ryan Florence (from React Router) has some good/interesting suggestions on app organisation/structuring that are worth considering. :ok_woman:

**Comment by levibuzolic** [comment]
Oops, forgot the link!

https://gist.github.com/ryanflorence/daafb1e3cb8ad740b346

**Comment by nii236** [comment]
@osiro: I think we should include propTypes in there, mostly because I don't use it, yet I should.

**Comment by osiro** [comment]
@levibuzolic Ryan's comments are good, I like it, as soon as the folder structure supports many non-related apps inside the same Rails app then it makes sense to. I just don't wanna have a top file named `App.jsx` because that makes no sense.

File: js-es2015-react.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+no-undef - disallow use of undeclared variables unless mentioned in a /*global */ block (recommended)
+no-unused-vars - disallow declaration of variables that are not used in the code (recommended)
+no-mixed-spaces-and-tabs - disallow mixed spaces and tabs for indentation (recommended)
+```
+
+The additional (not included in recommended) rules are:
+```
+"space-after-keywords": 1,
+"prefer-const": 1,
+"no-console": 1,
```
Comment (pzi) [inline js-es2015-react.md]: Same as L41

**Comment by levibuzolic** [comment]
> @levibuzolic Ryan's comments are good, I like it, as soon as the folder structure supports many non-related apps inside the same Rails app then it makes sense to. I just don't wanna have a top file named App.jsx because that makes no sense.

It totally does, just throw each app into a folder. His gist isn't a prescription, just a description, adapt it to fit your needs. :ok_hand: 

:wave:

File: js-es2015-react.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+# Sample Configuration
+
+Sample `.eslintignore` file:
+
+```
+node_modules
+```
+
+Sample `.eslintrc` file:
+
```
Comment (sheldongill) [inline js-es2015-react.md]: I'd prefer the sample files to be files. Just put a reference here.

File: js-es2015-react.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
+# Tooling
+
+For projects that use ES2015, we use:
+
+- eslint
```
Comment (pzi) [inline js-es2015-react.md]: Would be good to link those up:
- http://eslint.org/
- https://github.com/yannickcr/eslint-plugin-react

So newcomers/people can read up straight away

File: js-es2015-react.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+- Classes are PascalCased
+- Functions are camelCased
+- React Props and States are camelCased
+- Spaces, not tabs
+- Use propTypes where practicable
+- Promises are indented
+
+```
+Promise
+    .then(()=>{})
```
Comment (pzi) [inline js-es2015-react.md]: according to eslint rule, this should have 2 spaces only and some spaces could be added, too :smirk:

File: js-es2015-react.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+  }
+],
+"quotes": [2, "single"],
+"linebreak-style": [2, "unix"],
+"semi": [2, "always"]
+```
+
+## React Rules
+
+The following React rules used are:
```
Comment (pzi) [inline js-es2015-react.md]: Reads nicer imo: `The following React rules are used:`

**Comment by pzi** [comment]
# Syntax

I propose our component's methods to be laid out lifecycle methods first and render last:

**Basic component**:

```js
export default class NewComponent extends React.Component {
  constructor(props) {
    super(props);
  }
  componentWillMount() {}
  componentDidMount() {}
  componentWillReceiveProps(nextProps, nextContext) {}
  shouldComponentUpdate(nextProps, nextState, nextContext) {}
  componentWillUpdate(nextProps, nextState, nextContext) {}
  componentDidUpdate(prevProps, prevState, prevContext) {}
  componentWillUnmount() {}
  render() {}
}
```

**Component with custom methods:**

```js
export default class NewComponent extends React.Component {
  constructor(props) {
    super(props);
  }
  componentWillMount() {}
  componentDidMount() {}
  componentWillReceiveProps(nextProps, nextContext) {}
  shouldComponentUpdate(nextProps, nextState, nextContext) {}
  componentWillUpdate(nextProps, nextState, nextContext) {}
  componentDidUpdate(prevProps, prevState, prevContext) {}
  componentWillUnmount() {}

  /* custom methods */
  handleXyz() {}
  doSomething() {}

  render() {}
}
```

This way `render` is always to be found at the end of the file (the 'final', actual output) and everything that composed what is rendered/final, can be found above.

# Method names

Custom methods could be prefixed with `_` if people are so inclined?

Otherwise I'd say we
- name handlers `handleEventName`
- name handlers in props `onEventName`
- name methods that introduce more logic for the final `render` like `renderElement`.

# JSX as variable or return

Multiple lines should be wrapped in parentheses:

```js
const multiLineJsx = (
  <App>
    <Header />
    <Body />
    <Footer />
  </App>
);
```

Single lines can disregard the parentheses: `var singleLineJsx = <h1>Single Line</h1>;`

**Comment by nii236** [comment]
I think `_method` for private methods looks ugly and is unnecessary. 

I personally use `camelCase` for private methods and `PascalCase` for public methods. React apps usually contain only private methods, so I would use `handleClick()` for example.

File: README.md [hunk 8]
```diff
@@ -7,6 +7,20 @@ The standards to which code produced by The Frontier Group complies. Plus use pu
...
 * [RubyMotion](rubymotion.md)
 * [CSS](css.md)
-* [C#] (cSharp.md)
+* [C#](cSharp.md)
 * [HTML](html.md)
 * JavaScript (wanted)
+* [ES2015/JSX](js-es2015-react.md)
+
+# General
+The code editor shall have the following settings.
```
Comment (pzi) [inline README.md]: use `:` instead of `.`

File: README.md [hunk 9]
```diff
@@ -7,6 +7,20 @@ The standards to which code produced by The Frontier Group complies. Plus use pu
 * [Ruby](ruby.md)
 * [RubyMotion](rubymotion.md)
 * [CSS](css.md)
-* [C#] (cSharp.md)
+* [C#](cSharp.md)
 * [HTML](html.md)
 * JavaScript (wanted)
+* [ES2015/JSX](js-es2015-react.md)
```
Comment (pzi) [inline README.md]: Link needs updating

File: react.md [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+To plug into [eslint](http://eslint.org).
+
+# Styleguide
+
+General points:
+
+- React Props and States are camelCased
+- Use propTypes where practicable
+- Single line JSX to have parentheses
+- Multiline props when more than 5 props are in a component
```
Comment (pzi) [inline react.md]: For me, this needs explaining

File: react.md [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+- Multiline props to be in alphabetical order
+- Multiline props to be indented 2 spaces
+```
+<Component
+  AProp={this.handleAProp}
```
Comment (pzi) [inline react.md]: Should follow your rule on [L17](https://github.com/thefrontiergroup/standards/pull/21/files#diff-67f377003d8e128a03b229dc0a3f87a9R17)

File: react.md (hunk 10 above)
Comment (pzi) [inline react.md]: Personally, I try to fit props into the 100 characters per line and if it turns out that 1 has to go on a new line or it looks/feels weird I will put each of them on a new line.

**Pull request merged**